File: problems/003-memory-gate/solutions/python/main.py
import sys

t = list(map(int, sys.stdin.buffer.read().split()))
it = iter(t)
N, Q, C = next(it), next(it), next(it)
pi = [0] * (N + 1)
pm = [0] * (N + 1)
bad = [False] * (N + 2)
for i in range(1, N + 1):
    k, x, m = next(it), next(it), next(it)
    bad[i] = k == 64 or x > C or (m != -1 and m < x)
    pi[i] = pi[i - 1]
    pm[i] = pm[i - 1]
    if not bad[i]:
        pi[i] += x
        pm[i] += C if m == -1 else min(C, m)
nxt = [N + 1] * (N + 2)
for i in range(N, 0, -1):
    nxt[i] = i if bad[i] else nxt[i + 1]
out = []
for _ in range(Q):
    l, r = next(it), next(it)
    out.append(
        f"REJECT {nxt[l]}"
        if nxt[l] <= r
        else f"ACCEPT {(pi[r]-pi[l-1])*65536} {(pm[r]-pm[l-1])*65536}"
    )
print("\n".join(out))

File: problems/003-memory-gate/solutions/rust/main.rs
use std::io::{self, Read};
fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let mut it = s.split_whitespace();
    let n: usize = it.next().unwrap().parse().unwrap();
    let q: usize = it.next().unwrap().parse().unwrap();
    let c: u64 = it.next().unwrap().parse().unwrap();
    let mut pi = vec![0u64; n + 1];
    let mut pm = vec![0u64; n + 1];
    let mut bad = vec![false; n + 2];
    for i in 1..=n {
        let k: u32 = it.next().unwrap().parse().unwrap();
        let x: u64 = it.next().unwrap().parse().unwrap();
        let m: i64 = it.next().unwrap().parse().unwrap();
        bad[i] = k == 64 || x > c || (m >= 0 && (m as u64) < x);
        pi[i] = pi[i - 1];
        pm[i] = pm[i - 1];
        if !bad[i] {
            pi[i] += x;
            pm[i] += if m < 0 { c } else { c.min(m as u64) };
        }
    }
    let mut nxt = vec![n + 1; n + 2];
    for i in (1..=n).rev() {
        nxt[i] = if bad[i] { i } else { nxt[i + 1] };
    }
    let mut out = String::new();
    for _ in 0..q {
        let l: usize = it.next().unwrap().parse().unwrap();
        let r: usize = it.next().unwrap().parse().unwrap();
        if nxt[l] <= r {
            out += &format!("REJECT {}\n", nxt[l]);
        } else {
            out += &format!(
                "ACCEPT {} {}\n",
                (pi[r] - pi[l - 1]) * 65536,
                (pm[r] - pm[l - 1]) * 65536
            );
        }
    }
    print!("{}", out);
}
